# Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

## Instructions:

This challenge consists of the following subsections:

* Split the Data into Training and Testing Sets

* Create a Logistic Regression Model with the Original Data

* Predict a Logistic Regression Model with Resampled Training Data 

### Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

1. Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

2. Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

    > **Note** A value of `0` in the “loan_status” column means that the loan is healthy. A value of `1` means that the loan has a high risk of defaulting.  

3. Check the balance of the labels variable (`y`) by using the `value_counts` function.

4. Split the data into training and testing datasets by using `train_test_split`.

### Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

1. Fit a logistic regression model by using the training data (`X_train` and `y_train`).

2. Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.

4. Answer the following question: How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use `RandomOverSampler`.

To do so, complete the following steps:

1. Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

2. Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.
    
4. Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the `README.md` file included in your GitHub repository.

Structure your report by using the report template that `Starter_Code.zip` includes, and make sure that it contains the following:

1. An overview of the analysis: Explain the purpose of this analysis.


2. The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

3. A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import balanced_accuracy_score

import warnings
warnings.filterwarnings('ignore')

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df_lending_data = pd.read_csv(
    Path("Resources/lending_data.csv"),
    index_col="loan_size")

# Review the DataFrame
df_lending_data.head()
df_lending_data.tail()

interest_rate  borrower_income  debt_to_income  num_of_accounts  \
loan_size                                                                    
19100.0           11.261            86600        0.653580               12   
17700.0           10.662            80900        0.629172               11   
17600.0           10.595            80300        0.626401               11   
16300.0           10.068            75300        0.601594               10   
15600.0            9.742            72300        0.585062                9   

           derogatory_marks  total_debt  loan_status  
loan_size                                             
19100.0                   2       56600            1  
17700.0                   2       50900            1  
17600.0                   2       50300            1  
16300.0                   2       45300            1  
15600.0                   2       42300            1

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
target = df_lending_data["loan_status"]

# Separate the X variable, the features
features = df_lending_data.drop(columns="loan_status")

In [4]:
# Review the y variable Series
target.head()

loan_size
10700.0    0
8400.0     0
9000.0     0
10700.0    0
10800.0    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
features.head()

interest_rate  borrower_income  debt_to_income  num_of_accounts  \
loan_size                                                                    
10700.0            7.672            52800        0.431818                5   
8400.0             6.692            43600        0.311927                3   
9000.0             6.963            46100        0.349241                3   
10700.0            7.664            52700        0.430740                5   
10800.0            7.698            53000        0.433962                5   

           derogatory_marks  total_debt  
loan_size                                
10700.0                   1       22800  
8400.0                    0       13600  
9000.0                    0       16100  
10700.0                   1       22700  
10800.0                   1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
target.head()

loan_size
10700.0    0
8400.0     0
9000.0     0
10700.0    0
10800.0    0
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
training_features, testing_features, training_targets, testing_targets = train_test_split(features, target)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(training_features, training_targets)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
predictions = logistic_regression_model.predict(testing_features)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Print the balanced_accuracy score of the model
accuracy_score(testing_targets, predictions)

0.9914362360709864

In [11]:
# Generate a confusion matrix for the model
confusion_matrix(testing_targets, predictions)


array([[18647,    98],
       [   68,   571]], dtype=int64)

In [12]:
testing_targets

loan_size
9500.0     0
8700.0     0
9100.0     0
8500.0     0
11100.0    0
          ..
9000.0     0
10600.0    0
8600.0     0
8900.0     0
9200.0     0
Name: loan_status, Length: 19384, dtype: int64

In [13]:
predictions

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [14]:
# Print the classification report for the model
training_report = classification_report(testing_targets, predictions)

print(training_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18745
           1       0.85      0.89      0.87       639

    accuracy                           0.99     19384
   macro avg       0.92      0.94      0.93     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** For this test data, accuracy is extremely good as over 99.2% of the transactions were accurately categorized by the model. It also predicted true 18705 loans that were actually true and predicted false 103 loans that were actually true. Actually false that were predicted true were 66, and actually false that were predicted false were 578. If we compare the support number for actual high-risk loans (1) of 644 vs the actually false that were predicted false of 578, it seems to be quite good model as the error is very small (644 actual vs 578.) In addition, the F1 score is perfect at detecting the number of healthy loans that are accurately sampled (1) vs. the number of high-risk loans of 0.88%. The class of healthy loans (0) is extremely large compared to the class of high-risk loans (1). As such, given the imbalance, it means that the model is good at predicting the larger class, not the smaller high-risk loans' class that we would like to predict.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [15]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_oversampler = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
features_resampled, targets_resampled = random_oversampler.fit_resample(training_features, training_targets)

In [16]:
# Count the distinct values of the resampled labels data
targets_resampled.value_counts()

1    56291
0    56291
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [17]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
regression_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
regression_model.fit(features_resampled, targets_resampled)

# Make a prediction using the testing data
resampled_prediction = regression_model.predict(testing_features)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [19]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_score(testing_targets, resampled_prediction)

0.9931535211685036

In [21]:
# Generate a confusion matrix for the model
confusion_matrix(testing_targets, resampled_prediction)

array([[18635,   110],
       [    5,   634]], dtype=int64)

In [23]:
# Print the classification report for the model
training_report = classification_report(testing_targets, resampled_prediction)

print(training_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18745
           1       0.85      0.99      0.92       639

    accuracy                           0.99     19384
   macro avg       0.93      0.99      0.96     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The model fits a bit better the oversampled data, but only slightly when we review the F1 data of 88% vs 92% with the resampled data. All other metrics to predict accuracy were the same and, as such, we should use another method(s)to ascertain the accuracy of the model.